src/libstd/sync/mpsc/mod.rs

```diff
@@ -648,10 +648,18 @@ enum Flavor<T> {
 trait UnsafeFlavor<T> {
     fn inner_unsafe(&self) -> &UnsafeCell<Flavor<T>>;
     unsafe fn inner_mut(&self) -> &mut Flavor<T> {
-        &mut *self.inner_unsafe().get()
+        // SAFETY: We are sure the inner value will never be NUL when this is
+        // called, the invariants of the module make it so.
+        //
+        // It is also ensured that no other (mutable) reference will be handed
+        // out while the one returned here is in action.
```

Comment:
This seems like the wrong safety guarantee -- this function is unsafe, so it seems quite likely that the actually guarantee is that the caller has already checked these invariants for us. i.e., this function is just passing the proof obligations onto its callers.

Reply:
Updated to talk about the guarantees from the caller.